dpd-client/src/lib.rs

```diff
@@ -28,3 +29,106 @@ pub struct ClientState {
     /// Used for logging requests and responses.
     pub log: Logger,
 }
+
+impl Client {
+    /// Ensure that a NAT entry exists, overwriting a previous conflicting entry if
+    /// applicable.
+    #[allow(clippy::too_many_arguments)]
+    pub async fn ensure_nat_entry(
+        &self,
+        log: &Logger,
+        target_ip: ipnetwork::IpNetwork,
+        target_mac: types::MacAddr,
+        target_first_port: u16,
+        target_last_port: u16,
+        target_vni: u32,
+        sled_ip_address: &std::net::Ipv6Addr,
+    ) -> Result<(), progenitor_client::Error<types::Error>> {
+        let existing_nat = match target_ip {
+            ipnetwork::IpNetwork::V4(network) => {
+                self.nat_ipv4_get(&network.ip(), target_first_port).await
+            }
+            ipnetwork::IpNetwork::V6(network) => {
+                self.nat_ipv6_get(&network.ip(), target_first_port).await
+            }
+        };
+
+        // If a NAT entry already exists, but has the wrong internal
+        // IP address, delete the old entry before continuing (the
+        // DPD entry-creation API won't replace an existing entry).
+        // If the entry exists and has the right internal IP, there's
+        // no more work to do for this external IP.
+        match existing_nat {
+            Ok(existing) => {
+                let existing = existing.into_inner();
+                if existing.internal_ip != *sled_ip_address {
+                    info!(log, "deleting old nat entry";
+                      "target_ip" => ?target_ip);
+
+                    match target_ip {
+                        ipnetwork::IpNetwork::V4(network) => {
+                            self.nat_ipv4_delete(
```

Comment:
There is an opportunity for a race condition between these delete calls, and the calls to `nat_ipv[4,6]_create` below. Perhaps this is ok if we assume Nexus will never be operating on a single NAT entry concurrently (I don't know enough about Nexus to know if that is a reasonable assumption). I'm currently working on a [similar function](https://github.com/oxidecomputer/dendrite/pull/333) inside Dendrite itself to combine the delete/create stages of the ensure operation in a region guarded by a mutex.

Reply:
> Perhaps this is ok if we assume Nexus will never be operating on a single NAT entry concurrently (I don't know enough about Nexus to know if that is a reasonable assumption).

Nexus shouldn't be doing this, at least for iinstance NAT entries. During instance creation (IIUC) only one Nexus should be running the instance's create saga. If an instance migrates, Nexus won't allow a new migration to begin (which could kick off another parallel round of updates) until it's updated NAT entries to reflect the migration that just finished.